src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -4011,7 +4013,7 @@ private void ParseParameterModifiers(SyntaxListBuilder modifiers)
                 Debug.Assert(typeParameterList == null, "Exit condition of ParseMemberName in this scenario");
 
                 // No need for a diagnostic, ParseMemberName has already added one.
-                var missingIdentifier = CreateMissingIdentifierToken();
+                var missingIdentifier = CreateMissingIdentifierToken().WithAnnotationsGreen(new SyntaxAnnotation[] { new SyntaxAnnotation("value", discardedExplicitEventName) });
```

Comment:
this seems weird.  are annotations normally used for this purpose?

Reply:
is the goal to simply report a good error message here?  or is error message reported at some other location?

Reply:
The goal of this pull request is to fix the issue https://github.com/dotnet/roslyn/issues/19034. 

The issue, in short, is that the member name is missing in the error message in a certain case.

The member name is missing because the associated token is removed from the parse tree and replaced with a missing identifier token (that we are creating on the line), so the real name is gone.

In order to produce the error message correctly with the member name (without altering the token replacement logic), I used annotation. That allow me to pass some information so that it can be used to produce the error message.

That said, this use of annotation is unconventional.

Looking backward, the reason that we need to use annotation is because I avoided the token replacement logic. What if I actually do NOT replace the token? I have just done with the experiment and found those unit test failures.

Reply:
> The member name is missing because the associated token is removed from the parse tree and replaced with a missing identifier token (that we are creating on the line), so the real name is gone.

Diagnostics and Tokens are unrelated.  You can have a missing token that has a Diagnostic with rich information on it (like the 'real name' of the identifier).